### ISSUE TITLE
Device not showing up in linked devices

### ISSUE THREAD

**Issue by susanhita** [googlecodelabs/smarthome-washer#26]
I followed the codelabs, but on SYNC, there was a "Signing in" popup, and then nothing. The device did not show up in linked devices.

The last log says
"Function execution took 6 ms, finished with status code: 302"
the one before that says
"undefined?code=xxxxxx&state=undefined"

**Comment by Fleker** [comment]
So you got a sync request, sent back a response, but nothing happened?

**Comment by susanhita** [comment]
Yes nothing happened. There is nothing in linked devices.
![ezgif com-crop](https://user-images.githubusercontent.com/14327949/61579628-63eb1080-ab25-11e9-9358-3507dcd83ade.gif)

**Comment by Fleker** [comment]
Can you verify that you are sending the correct data in your sync response?

**Comment by jcartw89** [comment]
I seem to be having a similar issue when following the tutorial. At the end of Section 3 "Run the starter app" I launch the test for "On device testing". At which point I put `https://us-central1-washer-tutorial-121f7.cloudfunctions.net` (my project id is `washer-tutorial-121f7`) into the development server URL field. I then navigate to the "Add devices" page in Google Assistant and click `[test] washer tutorial`. However, no devices are added.

I can see in my firebase logs that the `https://us-central1-washer-tutorial-121f7.cloudfunctions.net/fakeauth` route is getting hit and is successfully doing a redirect to `https://oauth-redirect.googleusercontent.com/r/washer-tutorial-121f7?code=<CODE>&state=<STATE>`. However, it seems like the redirect is never successfully hitting the `https://us-central1-washer-tutorial-121f7.cloudfunctions.net/faketoken` route afterwards to complete the OAuth process. Any thoughts on what might be failing?

**Comment by Fleker** [comment]
This may be related to a similar authentication issue that we are investigating, although I do want to make sure you are returning the washer in your sync response.

**Comment by andrik78** [comment]
Hello @Fleker. Following the codelabs we are experiencing similar issue during device linking. In particular, no faketoken is required after (what seems to me) a positive fakeauth response. Could you plase advise on that? fakeauth response as follows:

https://oauth-redirect.googleusercontent.com/r/--------------?code=xxxxxx&state=AB8b_TMznnlTeYl_m3xrRwLVgpBv5C4-yVPo5NK8ES9MTzuhWxAPRFCLg8Ohq8_jta-hNXdqFIwBIrZDHsDwQ2elAGiP-vrE7BNegn7Pc5O4PvhxrwXiSkIPF8wIGw2M1g8zg5X-6ypAiUDa9muYav_gzmQ2hWnoNNhNF_NGIBX6O1EcaGnZfsIsnAlla9L351Fei5OvyWCchoFAGI9ekCg3s1zHAAmqdlJ-o-dcWmmZMtt-xFKKiZhJIYvnEFWg1k8tPOV014wpdxJ2tIhOZ6sxdlva8syJREP5QYF5IRzi8jFrX0K0gOX7xEp2svFIovzNJeLqHhfkI49GpUo5A_Ha5D6ERdh2GI4Lt7AkZoaoKCLI6FlxGQPaMmi_IXMUo-vymzY_mTwI5tFp0BGgnUGALJm71aod1NnrFRkkKqP7c8-c6LyzNo4n31gjSg_YKdBST4VkU3VSj3cC3ee-akTnc3vBAE3GO6Wc8mq-C2Vo-rPTkijHgCmjdIucA8OqbI3oChwpb-xLkLlSvFFZoqQjsyhuWXDLvdxOu9UAYvqka_iTZRJkbtE

**Comment by Fleker** [comment]
Can you try to create a new project and see if it works?

**Comment by Chris-mazo26** [comment]
Hi, I have the same problem as previous users, I have already done several projects and with different accounts and devices, the same error still appears, please help

**Comment by andrik78** [comment]
Hello @Fleker, following your suggestion I created several new projects (also with different users), always resulting in the authentication issue I reported before, i.e. no token request from smartphone. Could you please comment on that? Is something missing in the tutorial or something should be changed in the code which is not clearly specified? I thank you in advance for your support.

**Comment by kiralikbeyin** [comment]
Unlink - re Link account in test>simulator can work

**Comment by charlie-cheetah** [comment]
Hi, 
I have the same issue as above. I'm stuck on Section 3. I see in firebase logs only `/fakeauth` route is getting hit, and then Home App is getting back to "Add devices" screen.

**Comment by kiralikbeyin** [comment]
@charlie-cheetah Please can you try this:

from https://github.com/googlecodelabs/smarthome-washer/issues/28#issuecomment-529521608

5-Another ugly developer solution (yes it works like a first signup and links your devices) is below, slow down page transitions:

I found a way to usb debuging with chrome:
Check this first : #26 (comment)
GH goes to auth0.com and quickly returns back. **Before returning to "Add devices" screen if you hold your finger in a 100-500ms you will see "Return App" screen and it will stay on this page.** Open it with chrome. Maybe this can help someone..

**Comment by charlie-cheetah** [comment]
Thanks @kiralikbeyin I used auth0 instead of fakeauth and it works! The codelabs example should be updated though.

**Comment by kiralikbeyin** [comment]
@charlie-cheetah I have a problem with auth0.com
Everything is fine when using auth0.com standart dev keys which is NOT for production.

When you add your real keys which you will get from gcp https://auth0.com/docs/connections/social/google 
First login works fine. If you unlink and try again linking it doesnt work until you delete the user from auth0.com

Please if you can find a way stop autologin after unlink write here..

**Comment by Fleker** [comment]
What if your `/fakeauth` endpoint added a timeout? Would that provide more consistency? I wonder if the issue is due generally to some kind of race condition.

```
exports.fakeauth = functions.https.onRequest((request, response) => {
  const responseurl = util.format('%s?code=%s&state=%s',
    decodeURIComponent(request.query.redirect_uri), 'xxxxxx',
    request.query.state);
  console.log(responseurl);
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(response.redirect(responseurl));
    }, 1000)
  })
});
```

**Comment by javiercuellar73** [comment]
Same situation here. 
Like 1 month ago I was able to go through all the example. Algo change authentication to Auth0 and modify all functions to connect to my DB. I got stuck so I decide to start over with smarthome-washer example using a completly new project. I didn't modify anything from original index.js.
Logs shows only access to "fakeauth".  Should call "faketoken" but never doe. 
After that, I create a new project and same result.

Just test with the timeout, but same result.
Looks like this is something else out of our hands.

Any ideas?

**Comment by javiercuellar73** [comment]
> This may be related to a similar authentication issue that we are investigating, although I do want to make sure you are returning the washer in your sync response.

Hi Fleker
Were you able to find the similar authentication issue you had? 
Please let me know. Thanks!!!

**Comment by javiercuellar73** [comment]
Hi Fleker.
Any help or recomendation on this topic issue?  

Thanks!!!

**Comment by Fleker** [comment]
It's unclear what is going on. Are you sure your project is in a test state?

**Comment by javiercuellar73** [comment]
Yes it is. "Testing on Device: Enabled"
I didn't modify anything from original index.js.

**Comment by jcartw89** [comment]
@javiercuellar73 if you want to spin up a local auth server with Node.js and Ngrok, I've got a repo with all the OAuth code: https://github.com/jcartw89/google-your-smart-thing.

It doesn't implement true/secure OAuth, but it's good enough to get your smart home app POC working. Here's a guide on setting it all up: https://medium.com/swlh/ready-to-google-your-smart-thing-791964f2682f. 

You should be able to mash this together with the `smarthome-washer` tutorial if you're having still having issues.

**Comment by javiercuellar73** [comment]
@jcartw89   Thanks for the help.
I followed your tutorial. 
First I remove ALL projects from google cloud to have a clean environment.
Then I insert    
```
console.log("AUTHORIZE....");
console.log("LOGIN.....");
console.log("REFRESH TOKEN......");
```
at the begging of each module.exports.....   That way I can know when the function is hit.

After set up is completed, I go to Google Home from mobile phone, hit "[test] The Smart Thing Train"  (here the console shows the log AUTHORIZE....").
The login.html shows up. I click the button "Accept", the console shows the log "LOGIN...." and just after that the log "AUTHORIZE....."   (this is strange!!!!!).  The phone returns to "Manage Accounts" page.

It never hits the "/token" function, so still the same problem.

Any testing suggestion?

Thanks for your help!!!

**Comment by jcartw89** [comment]
My pleasure! So, based on that flow, it seems like you have both the `Authorization URL` and `Token URL` pointed to the same endpoint (e.g. https://d8a236bd.ngrok.io/authorize). Can you confirm that you set `Token URL` to the `/token` route? The URL should look like https://d8a236bd.ngrok.io/token.

The ngrok logs should show which routes are getting hit during the OAuth flow.

You can see the OAuth routes in the codebase here: https://github.com/jcartw89/google-your-smart-thing/blob/ac0a39722ee75bc746fb0e3e900a97136a736da9/server.js#L11.

**Comment by javiercuellar73** [comment]
I added another console.log to  "/login"

`console.log(`${redirect_uri}?code=${code}&state=${state}`);`

After returning from login.html it sends correct link

`https://oauth-redirect.googleusercontent.com/r/xxxxxxxxxx?code=code&state=AB8b_TO_lW0At0mYJfYMesvGTbxva0ISUhU1It_Cmb7wAL3Cdwe3kv2Qq302XTMqikzno-I7B3bFXlrGMXEvBrZxsMo8QWMjIHQpRGOVDjWXd1DaKLKY7QMbgEsUQuBQsKSQ36chcrNM9VszdQIEYOXOmLYVdlQWbwLjD9ACnGMHSxeYvWaxKnSaXYyAP_69VZrnFw67WQLvs1HUm2wZkt9w7B3CtaLUlydoAq-s3cxiawYqabi26y-gB1Xbl-dgP5puMjvD5f74MJN9XDC68DsLj1smL4WpNeMpmKqA54ICvrDu9OKOt8WXXQ2DbuhBDaBxQxYnFt6Lo2TmKifLWbEIeJDyDaEzEdrR1qmZRgjPPH7B2jQgIY9Lv8tlvENaIyGZ51nhsTwj1ZtMCXx8HSQVPIZLjF-9dlQazlU6py76qnk77TdbsrgaAcrzwpoBrfrqjiwtUAHLvNUklO4NiNMBtGZ68jwSedyvTdSV0CjA0kDzmw-n2_oqLu6jjwEJdcGvfBtEuLHX1B3zMGUob57y9iVgtKAIW3Tx4GwjoCKYSuIIKISyYWy9y4NuQYXGsLSqrneSYeC_2cXs-xg9Y5OhBRvWNZjrPg`

The strange thing is that after that, the "/authorize" function is hit again instead of "/token"

**Comment by javiercuellar73** [comment]
> My pleasure! So, based on that flow, it seems like you have both the `Authorization URL` and `Token URL` pointed to the same endpoint (e.g. https://d8a236bd.ngrok.io/authorize). Can you confirm that you set `Token URL` to the `/token` route? The URL should look like https://d8a236bd.ngrok.io/token.
> 
> The ngrok logs should show which routes are getting hit during the OAuth flow.
> 
> You can see the OAuth routes in the codebase here: https://github.com/jcartw89/google-your-smart-thing/blob/ac0a39722ee75bc746fb0e3e900a97136a736da9/server.js#L11.

I thought the same thing, and validate that URLs are correct. (..../authorize  and ../token)

**Comment by jcartw89** [comment]
![1_BdtZbsGEAozvtgrFGcBNIw](https://user-images.githubusercontent.com/11306493/69185747-0926ed00-0af6-11ea-99bb-569e67a95d27.png)

Do you have this fulfillment URL updated as well?

**Comment by jcartw89** [comment]
Using that OAuth/fulfillment server, the flow should ideally go:
1.  GET /authorize
2 . POST /login
3. POST /token
4. POST /fulfill

At that point the SYNC intent response should provide the device list to your Google Home Graph.

**Comment by jcartw89** [comment]
So, I just pull down the repo to do a fresh test and confirmed that it's working with ngrok. Is your OAuth server responding with a 302 as a result of POST /login? If the redirect is failing the GET /authorize may be a retry request.

```
HTTP Requests                                                                               
-------------                                                                               
                                                                                            
POST /fulfill                  200 OK                                                       
POST /fulfill                  200 OK                                                       
POST /token                    200 OK                                                       
POST /login                    302 Found                                                    
GET  /authorize                200 OK  
```

**Comment by javiercuellar73** [comment]
Hi @jcartw89 
Just tested as suggested, fresh test with ngrok and it works.
BTW, in my original oAuth settings (without ngrok)  the URL was
https://myserver/googlesmarthome/authorize
https://myserver/googlesmarthome/token

do you thing that having  /googlesmarthome/   is a problem?

I tested https://myserver/googlesmarthome/token using postman and the link is working, so it is available. Stange things

**Comment by javiercuellar73** [comment]
Just change my web parameters to use 
https://myserver.com/authorize  and
https://myserver.com/token
and still same problem.... /token never hitted. 

Others tests:
a) 
Authorization URL:   https://6a980497.ngrok.io/authorize
Token URL:    https://myserver.com/token
Result:    
Authorize :  200 Ok
Login:  302  Found
Token:    Not hit

b) 
Authorization URL:   https://myserver.com/authorize
Token URL:    https://6a980497.ngrok.io/token
Result:    
Works properly

Ideas??

**Comment by javiercuellar73** [comment]
I think the problem could be because myserver uses  Self-Signed certificate for https?
What do you think?  

@Fleker, please could you validate that?

**Comment by Fleker** [comment]
Why are you self signing your cert? Can you try using a cloud function, like in this sample?

**Comment by javiercuellar73** [comment]
I use a self signed cert because I wanted to use my own server. 
Will self signing cert be a problem?

Using ngrok works fine, I will try **again** your washing machine example and post results. 
I got to this post because the problem was this:
https://github.com/googlecodelabs/smarthome-washer/issues/26#issuecomment-542831674

**Comment by javiercuellar73** [comment]
> Will self signing cert be a problem?

I use Let's Encrypt and now I have a trusted certificate.
I can confirm that was the problem not being able to hit "/token"

Thanks @jcartw89 and @Fleker  for your help.

**Comment by tapitlrs** [comment]
To solve this issues, when in the linking interface, select 3 dots in the upper right corner and choose to open with a web browser, the link process will be successful. I think this is a Google Home app bug, hopefully resolved soon.

**Comment by vinhkhai95** [comment]
To solve this issue, while in the link interface, select 3 dots in the upper right corner and choose to open in a web browser, the linking process will be successful. I think this is a Google Home app bug, hopefully resolved soon.

**Comment by ChinthakaJay** [comment]
@vinhkhai95 Have no idea why. But this actually worked.

**Comment by devunwired** [comment]
We have observed this happening on some devices when there is no authorization UI presented, so we've added a fake login page to the codelab project to address this. Feel free to open a new issue if you are still experiencing the problem with the updated code.

Fixed with a1d4102f12e4e31aac0e550d5ad730fab6d70529
